// vue.min.js/library/connection/tcp.js
require ("core.min.js")
var lib = require ("core.min.js/src")
var $$$ = require (lib.package ("vue.min.js"))
const {define, echo} = lib
const {zero, one} = lib

define (module).export (function (request, response, next) {
	const vue = new $$$.vue ($$$.express.exe (), request, response)
	const cross = {origin: null, over: null}
	const application = function () {
		if (request.error) response.status (request.error.code).send (request.error.message)
		else vue.app (request, response, next)
		}
	const the = new setup (vue, application, request, response, next, cross)
	if (vue.setup ()) {
		vue.app.use ($$$.vue.process)
		if (cross.setup = true) vue.app.use ($$$.express.static.path ({static: vue.app.dir.client}))
		if (vue.app.use ($$$.express.cross ({origin: (cross.origin = vue.is.access.file ())}))) {
			if (request.cross.origin) if (lib.un (cross.origin)) return vue.app (request, response, next)
			if (request.method === lib.http.request.method.option) return vue.app (request, response, next)
			if (vue.fire.wall ()) return vue.app (request, response, next)
			if (request.cross.referer) if (vue.is.access.referer ()) vue.io = true
			request.cross.over = vue.is.access.process ()
			}
		if (lib.un (vue.io)) {
			if (lib.un (request.cross.origin))
			if (lib.un (request.cross.over)) vue.io = true
			else vue.io = null
			else {
				if (cross.origin) if (vue.is.token (request, response)) vue.io = true
				else request.error = {code: lib.http.header.status.error.forbidden, message: "token"}, vue.io = null
				else request.error = {code: lib.http.header.status.error.forbidden, message: "cross:origin"}, vue.io = null
				}
			}
		if (the.continue) the.io ()
		}
	else application ()
	})

class setup {
	constructor (vue, application, request, response, next, cross) {
		this.continue = true
		this.vue = vue
		this.application = application
		this.request = request
		this.response = response
		this.next = next
		this.cross = cross
		}
	io () { return setup.io (this.vue, this.application, this.request, this.response, this.next, this.cross) }
	process () { return setup.process (this.vue, this.application, this.request, this.response, this.next, this.cross) }
	}

setup.io = function (vue, application, request, response, next, cross) {
	if (vue.io) {
		if (vue.package) {
			const db = vue.data.base ("app").then (function (client) {
				vue.app.db = client.use (db.connection.name)
				if (vue.data.cache ()) setup.process (vue, application, request, response, next, cross)
				else vue.app.db.setup (function () { if (vue.data.create ()) setup.process (vue, application, request, response, next, cross) }, next)
				})
			db.catch (function (error) {
				var code, message = lib.http.header.status.code [code = lib.http.header.status.error.timeout]
				response.status (code).send (lib.to.string (error))
				})
			}
		else next ()
		}
	else application ()
	}

setup.process = function (vue, application, request, response, next, cross) {
	if (request.cross.over) {
		if ($$$.app.setting ["package"]) {
			if (vue.path.get ("/package.json")) vue.app.get (vue.path.get ("/package.json"), function (request, response, next) {
				var data = $$$.express.package.convert ($$$.app.package.get (), true)
				if ($$$.app.delay) lib.time.sleep (function () { response.json (data) }, $$$.app.delay)
				else response.json (data)
				})
			if (vue.path.get ("/cgi-bin/package.json")) vue.app.get (vue.path.get ("/cgi-bin/package.json"), function (request, response, next) {
				var data = $$$.express.package.convert (vue.app.package.get (), true)
				if ($$$.app.delay) lib.time.sleep (function () { response.json (data) }, $$$.app.delay)
				else response.json (data)
				})
			}
		if ($$$.app.setting ["db:path"]) {
			if (vue.path.get ("/cgi-bin/db/setup")) vue.app.get (vue.path.get ("/cgi-bin/db/setup"), function (request, response, next) {
				response.end ()
				})
			if (vue.path.get ("/cgi-bin/db/dump")) vue.app.get (vue.path.get ("/cgi-bin/db/dump"), function (request, response, next) {
				response.end ()
				})
			if (vue.path.get ("/cgi-bin/db/collection")) vue.app.get (vue.path.get ("/cgi-bin/db/collection"), function (request, response, next) {
				response.end ()
				})
			}
		}
	else {
		if (vue.path.get ("/package.json")) vue.app.get (vue.path.get ("/package.json"), function (request, response, next) {
			if (vue.router.slot.exist (request.path)) response.error ("error:handler")
			else response.error ()
			})
		if (vue.path.get ("/cgi-bin/package.json")) vue.app.get (vue.path.get ("/cgi-bin/package.json"), function (request, response, next) {
			if (vue.router.slot.exist (request.path)) response.error ("error:handler")
			else response.error ()
			})
		if (vue.path.get ("/cgi-bin/script")) vue.app.get (vue.path.get ("/cgi-bin/script").concat ("/*"), function (request, response, next) {
			if (vue.router.slot.exist (request.path)) response.error ("error:handler")
			else response.error ()
			})
		if (vue.path.get ("/cgi-bin/style.css")) vue.app.get (vue.path.get ("/cgi-bin/style.css"), function (request, response, next) {
			return response.css (vue.style.css ())
			if (request.query.cache === vue.app.session.get ("cache").id) response.css (vue.style.css ())
			else response.css ()
			})
		if (vue.path.get ("/cgi-bin/sitemap.xml")) vue.app.get (vue.path.get ("/cgi-bin/sitemap.xml"), function (request, response, next) {
			response.end ()
			})
		if ("setup") {
			vue.plugin.distribution ({overwrite: true})
			}
		if ("size_of") {
			vue.app.use (vue.path.get ("/static"), function (request, response, next) {
				var path = ("/static").concat (request.path)
				var file = lib.path.join (vue.app.dir.static, request.path)
				vue.size_of.push (path, file)
				next ()
				})
			vue.app.use ("/", function (request, response, next) {
				var path = request.path
				var file = lib.path.join (vue.app.dir.client, request.path)
				vue.size_of.push (path, file)
				next ()
				})
			}
		vue.app.use (vue.path.get ("/cgi-bin/plugin"), $$$.express.static (vue.app.dir.plugin))
		vue.app.use (vue.path.get ("/static"), $$$.express.static (vue.app.dir.static))
		vue.app.use ($$$.express.static (vue.app.dir.client))
		var the
		if (the = vue.require ()) {
			for (var i in the.router) vue.app [the.router [i].method] (the.router [i].url, $$$.vue.response.promise (vue, the.router [i]))
			for (var i in the.try) if (vue.exist) vue.app.all (the.try [i].url, $$$.vue.response.promise (vue, the.try [i]))
			}
		vue.app.all ("*", function (request, response, next) {
			if (vue.router.slot.exist (request.path)) response.error ("error:handler")
			else response.error ()
			})
		}
	application ()
	}
